low_level_platform/impl/src/lf_rp2040_support.c

```diff
@@ -202,9 +199,120 @@ int lf_enable_interrupts_nested() {
  */
 int _lf_single_threaded_notify_of_event() {
   // notify main sleep loop of event
-  sem_release(&_lf_sem_irq_event);
+  if (sem_release(&_lf_sem_irq_event)) {
+    return 0;
+  }
+  return 1;
+}
+
+#else // LF_SINGLE_THREADED
+
+#warning "Threaded runtime on RP2040 is still experimental"
+
+/**
+ * @brief Get the number of cores on the host machine.
+ */
+int lf_available_cores() { return 2; }
+
+static void* (*thread_1)(void*);
+static void* thread_1_args;
+static int num_create_threads_called = 0;
+static semaphore_t thread_1_done;
+static void* thread_1_return;
+
+#define MAGIC_THREAD1_ID 314159
+
+void core1_entry() {
+  thread_1_return = thread_1(thread_1_args);
+  sem_reset(&thread_1_done, 1);
+}
+
+int lf_thread_create(lf_thread_t* thread, void* (*lf_thread)(void*), void* arguments) {
+  // make sure this fn is only called once
+  if (num_create_threads_called != 0) {
+    return 1;
+  }
+  thread_1 = lf_thread;
+  thread_1_args = arguments;
+  num_create_threads_called += 1;
+  sem_init(&thread_1_done, 0, 1);
+  multicore_launch_core1(core1_entry);
+  *thread = MAGIC_THREAD1_ID;
+  return 0;
+}
+
+int lf_thread_join(lf_thread_t thread, void** thread_return) {
+  if (thread != MAGIC_THREAD1_ID) {
+    return 1;
+  }
+  sem_acquire_blocking(&thread_1_done);
+  // release in case join is called again
+  if (!sem_release(&thread_1_done)) {
+    // shouldn't be possible; lf_thread_join is only called from main thread
+    return 1;
+  }
+  if (thread_return) {
+    *thread_return = thread_1_return;
+  }
+  return 0;
+}
+
+int lf_mutex_init(lf_mutex_t* mutex) {
+  recursive_mutex_init(mutex);
+  return 0;
+}
+
+int lf_mutex_lock(lf_mutex_t* mutex) {
+  recursive_mutex_enter_blocking(mutex);
+  return 0;
+}
+
+int lf_mutex_unlock(lf_mutex_t* mutex) {
+  recursive_mutex_exit(mutex);
   return 0;
 }
-#endif // LF_SINGLE_THREADED
+
+// condition variables "notify" threads using a semaphore per core.
+// although there are only two cores, may not use just a single semaphore
+// as a cond_broadcast may be called from within an interrupt
+int lf_cond_init(lf_cond_t* cond, lf_mutex_t* mutex) {
+  for (int i = 0; i < NUM_CORES; i++) {
+    sem_init(&(cond->notifs[i]), 0, 1);
+  }
+  cond->mutex = mutex;
+  return 0;
+}
+
+int lf_cond_broadcast(lf_cond_t* cond) {
+  for (int i = 0; i < NUM_CORES; i++) {
+    sem_reset(&(cond->notifs[i]), 1);
+  }
+  return 0;
+}
+
+int lf_cond_signal(lf_cond_t* cond) {
+  return lf_cond_broadcast(cond); // spurious wakeups, but that's ok
```

Comment:
Yes, spurious wakeups would be good to avoid, but it would complicate the implementation significantly to try to avoid them (i'm imagining an array to keep track of which threads are waiting on the condition variable). I took a look at https://github.com/raspberrypi/pico-sdk/pull/1101 which seems like an ideal implementation, but it's very complex.

Typical usage is like
```
do {
  lf_cond_wait(cond);
} while (condition);
```
So spurious wakeups only cause an extra iteration of this loop, which doesn't have much affect performance-wise.

Reply:
The PR you linked to looked complicated yes, we solved it a lot easier in FlexPRET here https://github.com/pretis/flexpret/blob/master/sdk/lib/src/cond.c